core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java

```diff
@@ -2749,6 +2749,28 @@ private void checkLiteral2(String expression, String expected) {
     sql(query).ok(expected);
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-3876">[CALCITE-3876]
+   * RelToSqlConverter should not combine Projects when top Project contains
+   * window function referencing window function from bottom Project</a>. */
+  @Test public void testWindowOnWindowDoesNotCombineProjects() {
+    final String query = "SELECT ROW_NUMBER() OVER (ORDER BY rn)\n"
+        + "FROM (SELECT *,\n"
+        + "  ROW_NUMBER() OVER (ORDER BY \"product_id\") as rn\n"
```

Comment:
What about we do not have any options for `OVER `, there would then has no reference for underlying inputs.

Reply:
I think `ROW_NUMBER` must have `ORDER BY` in its window, and if so, this concern is moot.